Introduce delays in APF controller fights over FlowSchemaStatus

MikeSpreitzer opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide/first-contribution.md#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

**What type of PR is this?**
> Uncomment only one ` /kind <>` line, hit enter to put that in a new line, and remove leading whitespace from that line:
>
> /kind api-change
> /kind bug

/kind cleanup

> /kind deprecation
> /kind design
> /kind documentation
> /kind failing-test
> /kind feature
> /kind flake

**What this PR does / why we need it**:
This PR introduces a stochastic delay before the config consuming controller overwrites the referential integrity feedback in the Status of a FlowSchema.  This is to dampen any fights
that may break out between controllers of different revisions, which might have different
opinions about how to set that bit.

This PR also introduces an integration test that runs six such controllers, three aiming to set correct values and three aiming to set incorrect values.  The test lets the fight play out over 32 rounds (2 for warmup, 30 for evaluation), and evaluates whether the time/writes was within wide bounds.

The controller is generalized to support this test: its configuration now includes a name, a bit indicating whether to set the feedback incorrectly, a possibly-fake clock to read, and a bit indicating whether to suppress the regular reactions to notifications (the test takes over from the regular workqueue).

**Which issue(s) this PR fixes**:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
NONE
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs

```

MikeSpreitzer committed:
Introduce delays in APF controller fights over FlowSchemaStatus

Use very simple fixed logic avoid a rush to countermand an existing
setting.

Co-Authored-By: Min Kim <291271447@qq.com>

Various tweaks in response to recent review

MikeSpreitzer commented:
@kubernetes/sig-api-machinery-pr-reviews

MikeSpreitzer commented:
/cc @lavalamp 
/cc @deads2k 
/cc @yue9944882

lixiang233 commented:
/retest

MikeSpreitzer commented:
/retest

fedebongio commented:
/assign @lavalamp

MikeSpreitzer commented:
/test pull-kubernetes-e2e-kind
/test pull-kubernetes-e2e-kind-ipv6

yue9944882 commented on staging/src/k8s.io/apiserver/pkg/util/flowcontrol/apf_controller.go:
```diff
@@ -52,6 +55,8 @@ import (
 	fclistersv1a1 "k8s.io/client-go/listers/flowcontrol/v1alpha1"
 )
 
+const timeFmt = "2006-01-02T15:04:05.999"
```
i suppose it's only for logging, why not calling `Time.MarshalJSON` instead?

MikeSpreitzer replied:
Yes, this is only used in log messages emitted at V level 5 --- which is used in debugging but not production.

Note also that the values at hand are of type `time.Time` rather than kube's own wrapper type.

I could have used `time.RFC3339Nano` directly.  In fact, I did at first.  But the random precision wore out its welcome.  I did not need the endless repetitions of the time zone either.

MikeSpreitzer commented on staging/src/k8s.io/apiserver/pkg/util/flowcontrol/apf_controller.go:
```diff
@@ -256,27 +297,25 @@ func (cfgCtl *configController) processNextWorkItem() bool {
 	}
 	klog.Error(err)
-	return false
+	return triedWrites, didWrites, false, 0
```
Right, good catch.

MikeSpreitzer commented on staging/src/k8s.io/apiserver/pkg/util/flowcontrol/apf_controller.go:
```diff
@@ -316,28 +359,51 @@ type fsStatusUpdate struct {
 	var errs []error
+	fieldManager := "api-priority-and-fairness-config-consumer-v1"
+	if cfgCtl.invertDangle {
+		fieldManager += "x"
```
I am not sure I understand the question.  This adjustment is here only for the sake of simulating what would happen in a scenario with multiple kube-apiservers of different version, disagreeing on how to set the Status.  That will not naturally happen in an integration test of one version, so we have to fake up a difference.

MikeSpreitzer commented on staging/src/k8s.io/apiserver/pkg/util/flowcontrol/apf_controller.go:
```diff
@@ -418,7 +494,7 @@ func (meal *cfgMeal) digestFlowSchemasLocked(newFSs []*fctypesv1a1.FlowSchema) {
 		// TODO: consider not even trying if server is not handling
 		// requests yet.
-		meal.presyncFlowSchemaStatus(fs, !goodPriorityRef, fs.Spec.PriorityLevelConfiguration.Name)
+		meal.presyncFlowSchemaStatus(fs, meal.cfgCtl.invertDangle != !goodPriorityRef, fs.Spec.PriorityLevelConfiguration.Name)
```
By its nature, this test has to inject a difference somewhere.

MikeSpreitzer commented:
/test pull-kubernetes-conformance-kind-ga-only-parallel
/test pull-kubernetes-e2e-gce

MikeSpreitzer commented:
/test pull-kubernetes-files-remake

lavalamp commented:
/assign @rngy

MikeSpreitzer commented:
/test pull-kubernetes-e2e-gce-ubuntu-containerd

deads2k commented on staging/src/k8s.io/apiserver/pkg/util/flowcontrol/apf_controller.go:
```diff
@@ -84,7 +89,14 @@ type RequestDigest struct {
 type configController struct {
-	queueSetFactory fq.QueueSetFactory
+	name                string // varies in tests of fighting controllers
+	ignoreNotifications bool   // for testing
```
Having test branches is something both @lavalamp and I have reservations about.

We don't place test only flags like this in the controllers.  We either factor the controller take fake object input, manually force the notifications, or create a private function which can be replaced (the `sync` funcs in pkg/controllers).  The private function is different because the piece substituted doesn't even get compiled in during the build.

MikeSpreitzer replied:
Revised, no more `ignoreNotifications` in the controller.

deads2k commented on staging/src/k8s.io/apiserver/pkg/util/flowcontrol/apf_controller.go:
```diff
@@ -221,10 +255,14 @@ func (cfgCtl *configController) initializeConfigController(informerFactory kubei
 }
 
+func (cfgCtl *configController) WaitForCacheSync(stopCh <-chan struct{}) bool {
+	return cache.WaitForCacheSync(stopCh, cfgCtl.plInformerSynced, cfgCtl.fsInformerSynced)
```
nit: we don't normally have functions for one liners.

MikeSpreitzer replied:
This is a public method of `TestableInterface`.

deads2k commented on staging/src/k8s.io/apiserver/pkg/util/flowcontrol/apf_controller.go:
```diff
@@ -256,27 +297,25 @@ func (cfgCtl *configController) processNextWorkItem() bool {
-// objects.  It either succeeds and returns `true` or logs an error
-// and returns `false`.
-func (cfgCtl *configController) syncOne() bool {
+func (cfgCtl *configController) SyncOne() (bool, bool, bool, time.Duration) {
```
godoc telling me what the return values are.  something, something, retry, waitTime?

MikeSpreitzer replied:
Added.

deads2k commented on staging/src/k8s.io/apiserver/pkg/util/flowcontrol/apf_controller.go:
```diff
@@ -256,27 +297,25 @@ func (cfgCtl *configController) processNextWorkItem() bool {
 	}
 	klog.Error(err)
-	return false
+	return triedWrites, didWrites, true, 0
```
I don't  see usage of the triedWrites or didWrites in the product code.  We normally check writes in unit tests by inspecting the .Actions of the fake clients.

MikeSpreitzer replied:
These are for an integration test; the clients are not fakes.

deads2k commented on staging/src/k8s.io/apiserver/pkg/util/flowcontrol/apf_controller.go:
```diff
@@ -316,28 +359,51 @@ type fsStatusUpdate struct {
+	fsStatusUpdates, wait := cfgCtl.lockAndDigestConfigObjects(newPLs, newFSs)
+	var triedWrites, didWrites bool
 	var errs []error
+	fieldManager := "api-priority-and-fairness-config-consumer-v1"
```
Looking at usage below, it's not obvious to me how this will vary between releases.  do you want to assign this based on `version.Get()` somehow?  I could get behind making this value a field on `cfgCtl`.

MikeSpreitzer replied:
We discussed this, and decided we did not need to have a fully formed plan yet, it was sufficient to put _some_ sort of version mark in here so that we have something to vary later if and when we need to.

deads2k commented on staging/src/k8s.io/apiserver/pkg/util/flowcontrol/apf_controller.go:
```diff
@@ -316,28 +359,51 @@ type fsStatusUpdate struct {
+		klog.V(4).Infof("%s writing Condition %s to FlowSchema %s, which had ResourceVersion=%s, because its previous value was %s", cfgCtl.name, string(enc), fsu.flowSchema.Name, fsu.flowSchema.ResourceVersion, fcfmt.Fmt(fsu.oldValue))
+		triedWrites = true
+		fs2, err := cfgCtl.flowcontrolClient.FlowSchemas().Patch(context.TODO(), fsu.flowSchema.Name, apitypes.StrategicMergePatchType, []byte(fmt.Sprintf(`{"status": {"conditions": [ %s ] } }`, string(enc))), metav1.PatchOptions{FieldManager: fieldManager}, "status")
+		if err == nil {
```
nit: style is switch/case for more than two mutually exclusive.

MikeSpreitzer replied:
revised accordingly.

deads2k commented on staging/src/k8s.io/apiserver/pkg/util/flowcontrol/apf_controller.go:
```diff
@@ -316,28 +359,51 @@ type fsStatusUpdate struct {
 			errs = append(errs, errors.Wrap(err, fmt.Sprintf("failed to set a status.condition for FlowSchema %s", fsu.flowSchema.Name)))
 		}
 	}
 	if len(errs) == 0 {
```
NewAggregate already handles this case.

MikeSpreitzer replied:
Thanks, simplified.

deads2k commented on staging/src/k8s.io/apiserver/pkg/util/flowcontrol/apf_controller.go:
```diff
@@ -316,28 +359,51 @@ type fsStatusUpdate struct {
+}
+
+func (cfgCtl *configController) noteStatusUpdate(fs, fs2 *fctypesv1a1.FlowSchema) {
+	klog.V(5).Infof("%s at %s: successful update of FlowSchema %s from ResourceVersion %s to %s", cfgCtl.name, cfgCtl.clock.Now().Format(timeFmt), fs.Name, fs.ResourceVersion, fs2.ResourceVersion)
```
it's a little weird to output your own time.  Pretty sure klog already gives you time in the ouptut, doesn't it?

MikeSpreitzer replied:
When testing with a fake clock, the klog framing does not include the time that matters.

deads2k commented on staging/src/k8s.io/apiserver/pkg/util/flowcontrol/apf_controller.go:
```diff
@@ -359,10 +425,20 @@ func (cfgCtl *configController) lockAndDigestConfigObjects(newPLs []*fctypesv1a1
 
 	meal.finishQueueSetReconfigsLocked()
 
+	if cfgCtl.embargoTime != nil && len(meal.fsStatusUpdates) == 0 && !meal.suppressedWrites {
```
why does whether or not writes were suppressed impact the clearing of the embargo?

MikeSpreitzer replied:
Added a comment explaining what's going on here.

deads2k commented on staging/src/k8s.io/apiserver/pkg/util/flowcontrol/apf_controller.go:
```diff
@@ -359,10 +425,20 @@ func (cfgCtl *configController) lockAndDigestConfigObjects(newPLs []*fctypesv1a1
-	klog.V(5).Infof("Switched to new API Priority and Fairness configuration")
-	return meal.fsStatusUpdates
+	if meal.suppressedWrites {
+		delay := DurationMax((*cfgCtl.embargoTime).Sub(cfgCtl.clock.Now()), time.Millisecond)
```
is `delay` the same as `wait` above?  if so, naming consistency please.

MikeSpreitzer replied:
yes, renamed.

MikeSpreitzer commented:
/test pull-kubernetes-dependencies
/test pull-kubernetes-bazel-build

MikeSpreitzer commented:
The force-push to 198a5c3fe83 fixed an overlooked gofmt and rebased and squashed.  I was getting inaccurate CI complaints about using klog v1; hopefully the rebase will properly pick up the switch to v2 and end the madness.

MikeSpreitzer commented:
No, the rebase did not end the madness!  The bazel build failed, see https://prow.k8s.io/view/gcs/kubernetes-jenkins/pr-logs/pull/90541/pull-kubernetes-bazel-build/1281494087261753345 --- which says

```
 INFO: Call stack for the definition of repository 'com_coreos_etcd_amd64' which is a http_archive (rule definition at /bazel-scratch/.cache/bazel/_bazel_root/7989b31489f31aee54f32688da2f0120/external/bazel_tools/tools/build_defs/repo/http.bzl:296:16):
 - <builtin>
 - /home/prow/go/src/github.com/kubernetes/kubernetes/build/workspace.bzl:155:9
 - /home/prow/go/src/github.com/kubernetes/kubernetes/build/workspace.bzl:54:5
 - /home/prow/go/src/github.com/kubernetes/kubernetes/WORKSPACE:54:1
ERROR: /home/prow/go/src/github.com/kubernetes/kubernetes/test/integration/apiserver/flowcontrol/BUILD:3:1: no such package 'vendor/k8s.io/klog': BUILD file not found in any of the following directories. Add a BUILD file to a directory to mark it as a package.
 - /home/prow/go/src/github.com/kubernetes/kubernetes/vendor/k8s.io/klog and referenced by '//test/integration/apiserver/flowcontrol:go_default_test'
ERROR: Analysis of target '//test/integration/apiserver/flowcontrol:go_default_test' failed; build aborted: no such package 'vendor/k8s.io/klog': BUILD file not found in any of the following directories. Add a BUILD file to a directory to mark it as a package.
 - /home/prow/go/src/github.com/kubernetes/kubernetes/vendor/k8s.io/klog
INFO: Elapsed time: 33.811s
INFO: 0 processes.
FAILED: Build did NOT complete successfully (4579 packages loaded, 36201 targets configured) 
```

Now .../flowcontrol/BUILD and vendor/modules.txt reference v2 of klog.  But still there are complaints about klog.

MikeSpreitzer commented:
The force-push to ad160906a56 attempts to fix an import of v1 of klog.

MikeSpreitzer commented:
/retest

MikeSpreitzer commented:
The force-push to c9d3f69b86a is a rebase.

MikeSpreitzer commented:
And the force-push to 457b931f833 does the s/ctl/ctlr/ in the introduced file `test/integration/apiserver/flowcontrol/fight_test.go`.

MikeSpreitzer commented:
The force-push to 3bd9a5d69ed finishes merging changes into `test/integration/apiserver/flowcontrol/fight_test.go`.

MikeSpreitzer commented:
Why did the @k8s-ci-robot add the `triage/needs-information` label to this PR?

yue9944882 commented:
/retest

MikeSpreitzer commented:
/test pull-kubernetes-e2e-gce-ubuntu-containerd

MikeSpreitzer commented:
/retest

MikeSpreitzer commented:
The force-push to 6aa540a6c7c is a rebase.

MikeSpreitzer commented:
/retest

lavalamp commented on staging/src/k8s.io/apiserver/pkg/util/flowcontrol/apf_controller.go:
```diff
@@ -269,37 +300,46 @@ func (cfgCtlr *configController) processNextWorkItem() bool {
+// - didWrites bool indicating whether status udates were successful;
+// - needRetry bool indicating whether to requeue with exponential backoff;
+// - wait time.Duration indicating to requeue after a specific delay.
+func (cfgCtlr *configController) SyncOne() (bool, bool, bool, time.Duration) {
```
Please at least name the return parameters. But it's even better to make a "report" struct object and return that.

MikeSpreitzer replied:
Made a report structure.

lavalamp reviewed (commented):
Some comments mostly about not mixing test and production code.

MikeSpreitzer commented on staging/src/k8s.io/apiserver/pkg/util/flowcontrol/apf_controller.go:
```diff
@@ -160,71 +186,76 @@ func newTestableController(
-			oldPL := oldObj.(*fctypesv1a1.PriorityLevelConfiguration)
-			if !apiequality.Semantic.DeepEqual(oldPL.Spec, newPL.Spec) {
-				klog.V(7).Infof("Triggered API priority and fairness config reloading due to spec update of PLC %s", newPL.Name)
+	if !ignoreNotifications {
```
Actually, the fix is even simpler.  In the test, the informers are never even started!

MikeSpreitzer commented on staging/src/k8s.io/apiserver/pkg/util/flowcontrol/apf_controller.go:
```diff
@@ -84,9 +89,15 @@ type RequestDigest struct {
 
+	// Indicates whether to give incorrect referential integrity
+	// feedback; used for testing
+	invertDangle bool
```
changed

MikeSpreitzer commented:
The force-push to bb971cdc460 adds some missing comments on public items, updates bazel, and squashes.

MikeSpreitzer commented:
/priority important-soon

ahg-g commented on staging/src/k8s.io/apiserver/pkg/util/flowcontrol/apf_controller.go:
```diff
@@ -269,37 +314,59 @@ func (cfgCtlr *configController) processNextWorkItem() bool {
+		report := cfgCtlr.SyncOne()
+		switch {
+		case report.NeedRetry:
 			cfgCtlr.configQueue.AddRateLimited(obj)
-		} else {
+		case report.Wait > 0:
+			cfgCtlr.configQueue.AddAfter(obj, report.Wait)
```
Two questions:
1) an event will add a work item to the queue and that will force the sync again irrespective of whether or not the wait is done, right?
2) for our purposes, is there a difference between `AddRateLimited(obj)/AddAfter(obj)` and `AddRateLimited(0)/AddAfter(0)`

MikeSpreitzer replied:
1. If by "event" you mean a notification from an informer (i.e., a call to one of its `cache.ResourceEventHandlerFuncs`) then the answer is yes.  And this is true for both `AddRateLimited` and `AddAfter` (in fact, the implementation of `AddRateLimited` just calculates a delay and calls `AddAfter`).

2. No real difference.  `obj` will always be `0` in the current code, and breaking that would involve a lot of code changes.

ahg-g commented on staging/src/k8s.io/apiserver/pkg/util/flowcontrol/apf_controller.go:
```diff
@@ -339,28 +410,46 @@ type fsStatusUpdate struct {
+	fsStatusUpdates, wait := cfgCtlr.lockAndDigestConfigObjects(newPLs, newFSs)
+	var triedWrites, didWrites bool
 	var errs []error
+	// fieldManager := "api-priority-and-fairness-config-consumer-v1"
```
remove commented out code

MikeSpreitzer replied:
done

adtac commented on test/integration/apiserver/flowcontrol/fight_test.go:
```diff
@@ -0,0 +1,162 @@
+	const testN = 30
+	nextTime := clk.Now()
+	for j := 0; j < 2+testN; j++ {
+		time.Sleep(time.Millisecond * 150)
```
why do we need this? could this potentially become flaky in the future?

MikeSpreitzer replied:
This is to allow certain goroutines --- specifically, the work queue workers in the controllers --- a chance to do their thing.  Yes, in principle it could be flaky.  I do not recall whether CPU overloads are prevented during integration testing.  Does anybody here know?

This is by no means the only test that has this issue.

FYI, the desire to be able to have tests that do not rely on sleeping to get goroutines done is the reason https://github.com/kubernetes/apiserver/blob/master/pkg/util/flowcontrol/fairqueuing/testing/clock/event_clock.go exists.  You see that our QueueSet tests use that.  Unfortunately, using the event clocks for controllers would require some updates to the workqueue package (since workqueue can merge items).

lavalamp replied:
Would it be possible to poll until the controller is in the desired state? Otherwise I think this is very likely to be flaky. Yes, we have significant CPU contention in tests.

lavalamp replied:
Even running `go test -parallel=100 -count=1000 -run=ConfigConsumerFight` or whatever, although confidence building, often fails to be as harsh as the actual CI test environment.

MikeSpreitzer replied:
1. I should correct my comment above.  The sleep is to allow the notifications to come in and update the informer local caches; the work queues are deliberately cut out in this test.

2. So yes, I think there is a way to test whether the needed state has been reached.  First, in each iteration, keep track of the last (thankfully the SyncOne calls are serialized!) ResourceVersion returned from each object write.  Then replace the sleep with a test of whether a notification of that ResourceVersion of that object has been received for each controller.  The first iteration will require its own setup, which is do-able.

MikeSpreitzer commented on staging/src/k8s.io/apiserver/pkg/util/flowcontrol/apf_controller.go:
```diff
@@ -160,24 +194,24 @@ func newTestableController(
+	cfgCtlr.configQueue = workqueue.NewNamedRateLimitingQueue(workqueue.NewItemExponentialFailureRateLimiter(200*time.Millisecond, 8*time.Hour), "priority_and_fairness_config_queue")
+
 	// ensure the data structure reflects the mandatory config
 	cfgCtlr.lockAndDigestConfigObjects(nil, nil)
```
Yes, because lockAndDigestConfigObjects ensures some minimal behavior even when there are no API objects that configure it.  The KEP describes how there is certain behavior that happens even without any configuration objects.

MikeSpreitzer committed:
Fix typo in comment in apf_controller.go

Co-authored-by: ahg-g <40361897+ahg-g@users.noreply.github.com>

MikeSpreitzer committed:
Fix another comment typo in apf_controller.go

Co-authored-by: Adhityaa Chandrasekar <github@adtac.in>

MikeSpreitzer commented on staging/src/k8s.io/apiserver/pkg/util/flowcontrol/apf_controller.go:
```diff
@@ -382,10 +471,25 @@ func (cfgCtlr *configController) lockAndDigestConfigObjects(newPLs []*fctypesv1a
+	// change any FlowSchema's Status.  A desire to make a change can
+	// show up here as either a suppressed write or a non-suppressed
+	// write.
+	if cfgCtlr.embargoTime != nil && len(meal.fsStatusUpdates) == 0 && !meal.suppressedWrites {
```
(a) no, the first conjunct merely tests whether there is an embargo time set at all, not whether it has expired

(b) the comment expands on "no status update", explaining that there are two senses of that

MikeSpreitzer commented on staging/src/k8s.io/apiserver/pkg/util/flowcontrol/apf_controller.go:
```diff
@@ -382,10 +471,25 @@ func (cfgCtlr *configController) lockAndDigestConfigObjects(newPLs []*fctypesv1a
+	// cleared: when this controller decides it no longer wants to
+	// change any FlowSchema's Status.  A desire to make a change can
+	// show up here as either a suppressed write or a non-suppressed
+	// write.
```
It is true that `meal.suppressedWrites` implies `cfgCtlr.embargoTime != nil && now.Before(*cfgCtlr.embargoTime)`, but the converse is not true.  Do not forget https://github.com/MikeSpreitzer/kubernetes/blob/2dbf4e29dd6eaf88f9a4c927d279d98d8c83bab3/staging/src/k8s.io/apiserver/pkg/util/flowcontrol/apf_controller.go#L712 , nor the fact that FlowSchema objects can be deleted.

MikeSpreitzer committed:
Tidy up apf_controller a bit in response to review

MikeSpreitzer commented on staging/src/k8s.io/apiserver/pkg/util/flowcontrol/apf_controller.go:
```diff
@@ -339,28 +410,46 @@ type fsStatusUpdate struct {
 // cfgCtlr and writes its consequent new configState.
-func (cfgCtlr *configController) digestConfigObjects(newPLs []*fctypesv1a1.PriorityLevelConfiguration, newFSs []*fctypesv1a1.FlowSchema) error {
-	fsStatusUpdates := cfgCtlr.lockAndDigestConfigObjects(newPLs, newFSs)
+func (cfgCtlr *configController) digestConfigObjects(newPLs []*fctypesv1a1.PriorityLevelConfiguration, newFSs []*fctypesv1a1.FlowSchema) (bool, bool, time.Duration, error) {
```
done

MikeSpreitzer commented on staging/src/k8s.io/apiserver/pkg/util/flowcontrol/apf_controller.go:
```diff
@@ -339,28 +410,46 @@ type fsStatusUpdate struct {
-		if err != nil {
+		klog.V(4).Infof("%s writing Condition %s to FlowSchema %s, which had ResourceVersion=%s, because its previous value was %s", cfgCtlr.name, string(enc), fsu.flowSchema.Name, fsu.flowSchema.ResourceVersion, fcfmt.Fmt(fsu.oldValue))
+		triedWrites = true
+		fs2, err := cfgCtlr.flowcontrolClient.FlowSchemas().Patch(context.TODO(), fsu.flowSchema.Name, apitypes.StrategicMergePatchType, []byte(fmt.Sprintf(`{"status": {"conditions": [ %s ] } }`, string(enc))), metav1.PatchOptions{FieldManager: cfgCtlr.asFieldManager}, "status")
```
done

MikeSpreitzer commented:
@adtac and @ahg-g : thanks for the reviews.  I have responded; can you PTAL?

MikeSpreitzer commented:
The force-push to 276e233e7d9 fixes a copy-and-paste oversight.

MikeSpreitzer commented on test/integration/apiserver/flowcontrol/fight_test.go:
```diff
@@ -0,0 +1,162 @@
+			// The order of the following iteration is not deterministic,
+			// and that is good.
+			for invert, slice := range ctlrs {
+				visit(invert, i, slice[i])
```
You mean make `ctlrs` a parameter of `foreach`?  I do not see why that would be an improvement.  It adds complexity for the sake of use cases that do not exist in the current code.  Or perhaps I misunderstand the suggestion.

MikeSpreitzer commented on staging/src/k8s.io/apiserver/pkg/util/flowcontrol/apf_controller.go:
```diff
@@ -339,28 +415,48 @@ type fsStatusUpdate struct {
+			if apierrors.IsConflict(err) {
+				why = "Conflict"
+			}
+			klog.V(5).Infof("%s at %s: attempted update of FlowSchema %s from ResourceVersion %s failed because %s", cfgCtlr.name, cfgCtlr.clock.Now().Format(timeFmt), fsu.flowSchema.Name, fsu.flowSchema.ResourceVersion, why)
```
This is definitely not an error condition.  This line of code can be reached when everything is working as expected.

MikeSpreitzer commented on test/integration/apiserver/flowcontrol/fight_test.go:
```diff
@@ -0,0 +1,162 @@
+			countAtStart = writeCount
+		}
+		wait := time.Hour
+		foreach(func(invert bool, i int, ctlr utilfc.TestableInterface) {
```
The `foreach` function _visits_ the controllers serially.  There is a comment inside the definition of `foreach` acknowledging that map enumeration order is not fixed and asserting that this is good.  That gives us coverage of more scenarios.  The caller of `foreach` is not guaranteed what the order of visitation is, and should not care.

MikeSpreitzer commented on test/integration/apiserver/flowcontrol/fight_test.go:
```diff
@@ -0,0 +1,162 @@
+		dt := now.Sub(tStart).Seconds()
+		avgInterval := dt / float64(dCount)
+		expectedInterval := utilfc.MaxEmbargo.Seconds() / float64(size+1)
+		lowBound, highBound := expectedInterval*0.1, expectedInterval*10
```
Are you thinking of the 150ms at https://github.com/MikeSpreitzer/kubernetes/blob/276e233e7d9b24a64eea942d2e03e75c16c5f321/test/integration/apiserver/flowcontrol/fight_test.go#L115 ?  Note that this is paid once for each of the 30 rounds.
Note that `expectedInterval` is 1/4 of 10 seconds --- that's 2.5 seconds.  One tenth of that is 250ms.

MikeSpreitzer commented:
/test pull-kubernetes-verify
/test pull-kubernetes-typecheck

MikeSpreitzer commented:
/retest

MikeSpreitzer commented:
The force-push to b2d789cac5a is a rebase.

MikeSpreitzer commented:
The force-push to 12a4f038ea6 fixes an oversight in the rebase.

MikeSpreitzer commented:
The force-push to d062241cd22 fixes another oversight.

MikeSpreitzer commented:
/test pull-kubernetes-bazel-test

lavalamp commented:
I really want to approve this but I'm worried about that time.Sleep...

MikeSpreitzer committed:
Prep for better wait, introduce TestableConfig

MikeSpreitzer commented on staging/src/k8s.io/apiserver/pkg/util/flowcontrol/apf_controller.go:
```diff
@@ -152,6 +181,11 @@ type priorityLevelState struct {
+	clock clock.PassiveClock,
+	finishHandlingNotification func(workqueue.RateLimitingInterface),
+	asFieldManager string,
+	foundToDangling func(bool) bool,
```
done

MikeSpreitzer commented:
/retest

MikeSpreitzer committed:
Make APF controller fight test clocked by notifications

MikeSpreitzer committed:
limit wait

MikeSpreitzer committed:
more log

MikeSpreitzer committed:
More log, trim junk, update bazel

MikeSpreitzer committed:
Fail if some time out

Also pick some lint.

MikeSpreitzer commented:
I need help here.  I have modified the logic in the fight test to wait for the exact needed condition rather than sleep 150 milliseconds, but am suffering two surprises:
1. The initial GETs are returning objects whose ResourceVersion is the empty string!
2. The informers in the six test controllers deliver notifications only during the initial sync, never after the writes during the fights!

You can see this in the pull-kubernetes-integration output, from file k8s.io/kubernetes/test/integration/apiserver/flowcontrol func TestConfigConsumerFight.  For surprise 1, see lines like
```
    fight_test.go:91: Got initial FlowSchema &v1beta1.FlowSchema{TypeMeta:v1.TypeMeta{Kind:"", APIVersion:""}, ObjectMeta:v1.ObjectMeta{Name:"catch-all", GenerateName:"", Namespace:"", SelfLink:"", UID:"", ResourceVersion:"", Generation:0, CreationTimestamp:v1.Time{Time:time.Time{wall:0x0, ext:0, loc:(*time.Location)(nil)}}, DeletionTimestamp:(*v1.Time)(nil), DeletionGracePeriodSeconds:(*int64)(nil), Labels:map[string]string(nil), Annotations:map[string]string(nil), OwnerReferences:[]v1.OwnerReference(nil), Finalizers:[]string(nil), ClusterName:"", ManagedFields:[]v1.ManagedFieldsEntry(nil)}, Spec:v1beta1.FlowSchemaSpec{PriorityLevelConfiguration:v1beta1.PriorityLevelConfigurationReference{Name:"catch-all"}, MatchingPrecedence:10000, DistinguisherMethod:(*v1beta1.FlowDistinguisherMethod)(0xc0004d3c80), Rules:[]v1beta1.PolicyRulesWithSubjects{v1beta1.PolicyRulesWithSubjects{Subjects:[]v1beta1.Subject{v1beta1.Subject{Kind:"Group", User:(*v1beta1.UserSubject)(nil), Group:(*v1beta1.GroupSubject)(0xc0004d3c00), ServiceAccount:(*v1beta1.ServiceAccountSubject)(nil)}, v1beta1.Subject{Kind:"Group", User:(*v1beta1.UserSubject)(nil), Group:(*v1beta1.GroupSubject)(0xc0004d3c10), ServiceAccount:(*v1beta1.ServiceAccountSubject)(nil)}}, ResourceRules:[]v1beta1.ResourcePolicyRule{v1beta1.ResourcePolicyRule{Verbs:[]string{"*"}, APIGroups:[]string{"*"}, Resources:[]string{"*"}, ClusterScope:true, Namespaces:[]string{"*"}}}, NonResourceRules:[]v1beta1.NonResourcePolicyRule{v1beta1.NonResourcePolicyRule{Verbs:[]string{"*"}, NonResourceURLs:[]string{"*"}}}}}}, Status:v1beta1.FlowSchemaStatus{Conditions:[]v1beta1.FlowSchemaCondition(nil)}}
```

For surprise 2, note that the last line like
```    fight_test.go:113: For invert=true, i=2, notified of &v1beta1.FlowSchema{TypeMeta:v1.TypeMeta{Kind:"", APIVersion:""}, ObjectMeta:v1.ObjectMeta{Name:"service-accounts", GenerateName:"", Namespace:"", SelfLink:"", UID:"ba222375-ea5a-426c-9f2e-5506e3c98664", ResourceVersion:"1784", Generation:1, CreationTimestamp:v1.Time{Time:time.Time{wall:0x0, ext:63741478626, loc:(*time.Location)(0x706cc20)}}, DeletionTimestamp:(*v1.Time)(nil), DeletionGracePeriodSeconds:(*int64)(nil), Labels:map[string]string(nil), Annotations:map[string]string(nil), OwnerReferences:[]v1.OwnerReference(nil), Finalizers:[]string(nil), ClusterName:"", ManagedFields:[]v1.ManagedFieldsEntry{v1.ManagedFieldsEntry{Manager:"api-priority-and-fairness-config-producer-v1", Operation:"Update", APIVersion:"flowcontrol.apiserver.k8s.io/v1beta1", Time:(*v1.Time)(0xc00b43aae0), FieldsType:"FieldsV1", FieldsV1:(*v1.FieldsV1)(0xc00b43ab00)}}}, Spec:v1beta1.FlowSchemaSpec{PriorityLevelConfiguration:v1beta1.PriorityLevelConfigurationReference{Name:"workload-low"}, MatchingPrecedence:9000, DistinguisherMethod:(*v1beta1.FlowDistinguisherMethod)(0xc00b53a630), Rules:[]v1beta1.PolicyRulesWithSubjects{v1beta1.PolicyRulesWithSubjects{Subjects:[]v1beta1.Subject{v1beta1.Subject{Kind:"Group", User:(*v1beta1.UserSubject)(nil), Group:(*v1beta1.GroupSubject)(0xc00b53a640), ServiceAccount:(*v1beta1.ServiceAccountSubject)(nil)}}, ResourceRules:[]v1beta1.ResourcePolicyRule{v1beta1.ResourcePolicyRule{Verbs:[]string{"*"}, APIGroups:[]string{"*"}, Resources:[]string{"*"}, ClusterScope:true, Namespaces:[]string{"*"}}}, NonResourceRules:[]v1beta1.NonResourcePolicyRule{v1beta1.NonResourcePolicyRule{Verbs:[]string{"*"}, NonResourceURLs:[]string{"*"}}}}}}, Status:v1beta1.FlowSchemaStatus{Conditions:[]v1beta1.FlowSchemaCondition(nil)}}
```
precedes
```
    fight_test.go:133: Achieved initial sync for invert=true, i=2
```

MikeSpreitzer committed:
wait longer for notifications

MikeSpreitzer commented:
And no, the longer wait does not solve the problem.

deads2k commented on test/integration/apiserver/flowcontrol/fight_test.go:
```diff
@@ -0,0 +1,219 @@
+					if err != nil {
+						return false, err
+					}
+					t.Logf("Got initial FlowSchema %#+v", fs)
```
This log line is printing your stub resource instance, not the resource instance that you got on line 87, which is why metadata is empty.

MikeSpreitzer replied:
fixed

deads2k commented on test/integration/apiserver/flowcontrol/fight_test.go:
```diff
@@ -0,0 +1,219 @@
+		if lastRVs == nil {
+			lastRVs = make(map[string]string)
+			// Wait until every FlowSchema is defined, and record its RV
+			allFlowSchemas := append(fcboot.MandatoryFlowSchemas, fcboot.SuggestedFlowSchemas...)
```
Based on usage, building a slice of names will make the problem you just had impossible to have again.  I'm unopinionated about doing that, just pointing out that the slice of strings wouldn't print like a real resource instance.

MikeSpreitzer replied:
More trouble that it is worth, IMO.

deads2k commented on test/integration/apiserver/flowcontrol/fight_test.go:
```diff
@@ -0,0 +1,219 @@
+		ctlrs[invert][i] = ctlr
+		informerFactory.Start(stopCh)
+		if ctlr.WaitForCacheSync(stopCh) {
+			t.Logf("Achieved initial sync for invert=%v, i=%v", invert, i)
```
You expressed surprise that this can print "late".  Since waiting is done for a short period and we assume it doesn't really matter how precise the handling is, it's a simple polling implementation, so it's not surprising that sometimes this print statement loses the race and you get a print from the controller first.

MikeSpreitzer replied:
The surprise is that there are _no_ print statements from the test notification handlers after the controllers start making changes.

MikeSpreitzer replied:
Argh, this should have been no surprise; the notification handler was reacting only if the spec changed.

deads2k commented on test/integration/apiserver/flowcontrol/fight_test.go:
```diff
@@ -0,0 +1,219 @@
+			AOK = true
+			foreach(func(invert bool, i int) {
+				for key, rv := range lastRVs {
+					if notifiedRVs[invert][i][key] != rv {
```
With this written on line 118, how do you avoid the data race?

MikeSpreitzer replied:
Good point, the writes are in different goroutines.

MikeSpreitzer replied:
Added synchronization.

deads2k commented on test/integration/apiserver/flowcontrol/fight_test.go:
```diff
@@ -0,0 +1,219 @@
+		clk.SetTime(now)
+		t.Logf("Syncing[size=%d, j=%d] at %s", size, j, now.Format(timeFmt))
+		klog.V(3).Infof("Syncing size=%d, j=%d at %s", size, j, now.Format(timeFmt))
+		if j == 2 {
```
why is the third run special?

MikeSpreitzer replied:
It's the first two that are special, the next 30 are ordinary.
We burn the first two to give it a better chance to get into an ordinary rhythm.

deads2k replied:
> It's the first two that are special, the rest are ordinary.
> We burn the first two to give it a better chance to get into an ordinary rhythm.

Can I get that as a comment in the code and maybe an explanation of why they're special in that comment.

MikeSpreitzer committed:
Patches to fight test func

Log the fetched initial object, not its precursor.

Add mutex around access to notifiedRVs because it is accessed from
various goroutines.

MikeSpreitzer committed:
Changes to FlowSchemaStatus matter

MikeSpreitzer commented:
/retest

deads2k commented on staging/src/k8s.io/apiserver/pkg/util/flowcontrol/apf_controller.go:
```diff
@@ -151,34 +179,27 @@ type priorityLevelState struct {
+		priorityLevelStates:        make(map[string]*priorityLevelState),
+	}
+	klog.V(2).Infof("NewTestableController %q with serverConcurrencyLimit=%d, requestWaitLimit=%s", cfgCtlr.name, cfgCtlr.serverConcurrencyLimit, cfgCtlr.requestWaitLimit)
+	cfgCtlr.configQueue = workqueue.NewNamedRateLimitingQueue(workqueue.NewItemExponentialFailureRateLimiter(200*time.Millisecond, 8*time.Hour), "priority_and_fairness_config_queue")
```
if you're going non-standard, please comment with why the different times are more suitable.

deads2k commented on staging/src/k8s.io/apiserver/pkg/util/flowcontrol/apf_controller.go:
```diff
@@ -189,42 +210,54 @@ func (cfgCtlr *configController) initializeConfigController(informerFactory kube
+
+// FinishHandlingNotification is the normal way to do so.
+// Pass this to NewTestable(..) to make a normal Interface.
+func FinishHandlingNotification(wq workqueue.RateLimitingInterface, obj interface{}) {
```
Didn't expect to see this exposed.  Also, this is just `AddZeroItemToQueue`?

deads2k commented on test/integration/apiserver/flowcontrol/fight_test.go:
```diff
@@ -0,0 +1,237 @@
+		ctlr := utilfc.NewTestable(utilfc.TestableConfig{
+			Name:  fmt.Sprintf("Controller%d[invert=%v]", i, invert),
+			Clock: clk,
+			FinishHandlingNotification: func(wq workqueue.RateLimitingInterface, obj interface{}) {
```
this doesn't add to the workqueue like the normal one?  That seems a little odd.

deads2k commented on staging/src/k8s.io/apiserver/pkg/util/flowcontrol/apf_controller.go:
```diff
@@ -247,7 +280,7 @@ func (cfgCtlr *configController) updateObservations() {
 	defer cfgCtlr.configQueue.ShutDown()
 	klog.Info("Starting API Priority and Fairness config controller")
-	if ok := cache.WaitForCacheSync(stopCh, cfgCtlr.plInformerSynced, cfgCtlr.fsInformerSynced); !ok {
+	if ok := cfgCtlr.WaitForCacheSync(stopCh); !ok {
```
why do we no longer wait for the caches to sync?

deads2k commented on staging/src/k8s.io/apiserver/pkg/util/flowcontrol/apf_controller.go:
```diff
@@ -247,7 +280,7 @@ func (cfgCtlr *configController) updateObservations() {
 	defer cfgCtlr.configQueue.ShutDown()
 	klog.Info("Starting API Priority and Fairness config controller")
-	if ok := cache.WaitForCacheSync(stopCh, cfgCtlr.plInformerSynced, cfgCtlr.fsInformerSynced); !ok {
+	if ok := cfgCtlr.WaitForCacheSync(stopCh); !ok {
```
This is usually expressed inline. Let's be consistent with other controllers and put the standard if-block back here.

MikeSpreitzer replied:
No, the fight test does _not_ invoke `Run`, but it needs the wait-for-cache-sync behavior.

deads2k commented on staging/src/k8s.io/apiserver/pkg/util/flowcontrol/apf_controller.go:
```diff
@@ -269,37 +302,64 @@ func (cfgCtlr *configController) processNextWorkItem() bool {
 	}
 	klog.Error(err)
-	return false
+	report.NeedRetry, report.Wait = true, 0
```
I'm not used to this construct.  What does the `0` mean?

MikeSpreitzer commented:
This was superseded by #99935
